src/Lean/Widget/InteractiveGoals.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Wojciech Nawrocki
+-/
+import Lean.Widget.ExprWithCtx
+
+/-! RPC procedures for retrieving tactic and term goals with embedded `ExprText`s. -/
+-- TODO(WN): this module is mostly a slightly adapted copy of the corresponding `plainGoal/plainTemGoal` handlers
+-- unify them somehow? or delete the plain ones.
+
+namespace Lean.Widget
+open Server
+
+structure InteractiveGoal where
+  hyps      : Array (String × ExprText)
```

Comment:
Just so that I understand this correctly: when the client gets an interactive goal, it also gets ownership of all the `LazyExprWithCtx` references in the goal.  These references need to be released every time the goal is refreshed.  Isn't that a lot of requests?

Reply:
It's a lot of requests and I have an outstanding TODO to batch them into a `Release where refs : Array RpcRef`. I'll add that.